### PR TITLE
Test execution using labels

### PR DESCRIPTION
**Pull request by k15r**
**DEADLINE:  Friday, March 27th 2020**

**Description**

Changes proposed in this pull request:

- Proposal on how to organize kyma tests using labels

**Related issue(s)**

### CODE REVIEW

**Comment by mareknawa** [comment]
I would love to see a section with bullet points what needs to be changed to go with this solution. As we discussed on SIG-Community Octopus has this functionality and effort to introduce this change is fairly small.

File: collaboration/sig-core/proposals/test-execution-using-labels.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
+# Test execution organization using labels
+
+Created 2020-03-16 by Korbinian Stoemmer (@k15r)
+
+## Motivation
+
+Kyma developers write a lot of tests. These tests have different intentions and thus require different conditions to run successfully. In order to run those tests under the expected conditions on our CI Infrastructure, multiple shell scripts have been implemented.
```
Comment (majakurcius) [inline collaboration/sig-core/proposals/test-execution-using-labels.md]:
```suggestion
Kyma developers write a lot of tests. These tests have different purposes and thus require different conditions to run successfully. In order to run those tests under the expected conditions on our CI Infrastructure, multiple shell scripts have been implemented.
```

File: collaboration/sig-core/proposals/test-execution-using-labels.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+Created 2020-03-16 by Korbinian Stoemmer (@k15r)
+
+## Motivation
+
+Kyma developers write a lot of tests. These tests have different intentions and thus require different conditions to run successfully. In order to run those tests under the expected conditions on our CI Infrastructure, multiple shell scripts have been implemented. 
+These shell scripts are tailored to the exact implementation of the test at the time of writing. This makes it really hard to reimplement a test, or to add a new test that requires the same setup (conditions) as another test.
+
+Example:
+
+currently the backup and restore tests are executed the following way on our CI:
```
Comment (majakurcius) [inline collaboration/sig-core/proposals/test-execution-using-labels.md]:
```suggestion
For example, currently the backup and restore tests are executed the following way on our CI:
```

File: collaboration/sig-core/proposals/test-execution-using-labels.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+* install kyma
+* install `TestDefinition` with parameter indicating before backup test
+* find all `TestDefinitions` with specific app label
+* create `ClusterTestSuite` referencing the found `TestDefinitions` and wait for success
+* take backup
+* spinup new cluster and restore
+* install `Testdefinition` with parameter indicating after restore test
+* create `ClusterTestSuite` referencing the found `TestDefinitions` and wait for success
```
Comment (majakurcius) [inline collaboration/sig-core/proposals/test-execution-using-labels.md]:
```suggestion
1. Install Kyma.
2. Install `TestDefinition` with parameter indicating before backup test.
3. Find all `TestDefinitions` with specific app label.
4. Create `ClusterTestSuite` referencing the found `TestDefinitions` and wait for success.
5. Take backup.
6. Spin up new cluster and restore.
7. Install `TestDefinition` with parameter indicating after restore test.
8. Create `ClusterTestSuite` referencing the found `TestDefinitions` and wait for success.
```

As per [guidelines](https://github.com/kyma-project/community/blob/master/guidelines/content-guidelines/04-formatting.md#ordered-and-unordered-lists), we use ordered lists if the order of the performed steps matter. They also state to capitalize the first letter of a bullet point, and end it with a period if it is a sentence.

File: collaboration/sig-core/proposals/test-execution-using-labels.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+
+For upgrade tests the situation is as follows:
+
+* install old kyma
+* install upgrade test `helm-chart`
+  * this creates a pod that does all the preparation work (compared to the backup test, this one does not use a pre-upgrade `TestDefinition`)
+* upgrade kyma
+* find **all** `TestDefinitions` in cluster
+* create `ClusterTestSuite` referencing the found `TestDefinitions` and wait for success
```
Comment (majakurcius) [inline collaboration/sig-core/proposals/test-execution-using-labels.md]:
```suggestion
1. Install old Kyma
2. Install upgrade test `helm-chart`. This creates a pod that does all the preparation work (compared to the backup test, this one does not use a pre-upgrade `TestDefinition`).
3. Upgrade Kyma.
4. Find **all** `TestDefinitions` in cluster.
5. Create `ClusterTestSuite` referencing the found `TestDefinitions` and wait for success.
```
As per [guidelines](https://github.com/kyma-project/community/blob/master/guidelines/content-guidelines/04-formatting.md#ordered-and-unordered-lists), we use ordered lists if the order of the performed steps matter. They also state to capitalize the first letter of a bullet point, and end it with a period if it is a sentence.

File: collaboration/sig-core/proposals/test-execution-using-labels.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+Backup/Restore and upgrade tests implement their own framework for executing multiple tests.Event though both tests should be very similar and support reusing test cases it is very hard to do so. This leads to tests that are implemented multiple times.
```
Comment (majakurcius) [inline collaboration/sig-core/proposals/test-execution-using-labels.md]:
```suggestion
Backup/Restore and upgrade tests implement their own framework for executing multiple tests. Event though both tests should be very similar and support reusing test cases, it is very hard to do so. This leads to tests being implemented multiple times.
```

File: collaboration/sig-core/proposals/test-execution-using-labels.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+## Goal
+
+* It should be easy to reuse tests for multiple scenarios.
+* It should be easy to implement tests without having to use yet another test runner. Octopus should be the only external requirement.
+* It should be possible execute an unlimited number of test-cases per scenario
```
Comment (majakurcius) [inline collaboration/sig-core/proposals/test-execution-using-labels.md]:
```suggestion
* It should be possible to execute an unlimited number of test cases per scenario.
```

File: collaboration/sig-core/proposals/test-execution-using-labels.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+* There should be one approach to run those tests from within the shell scripts. This reduces code in the shell scripts to scenario setup code.
+
+## Suggested solution
+
+Octopus allows to specify label queries for `ClusterTestSuites`. A defined set of labels should indicate the which `TestDefinition` to execute for which scenario.
```
Comment (majakurcius) [inline collaboration/sig-core/proposals/test-execution-using-labels.md]:
```suggestion
Octopus allows to specify label queries for `ClusterTestSuites`. A defined set of labels should indicate which `TestDefinition` to execute for which scenario.
```

File: collaboration/sig-core/proposals/test-execution-using-labels.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+###The labels:
```
Comment (majakurcius) [inline collaboration/sig-core/proposals/test-execution-using-labels.md]:
```suggestion
### The labels
```

File: collaboration/sig-core/proposals/test-execution-using-labels.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+each `TestDefinition` should be labeled with one or multiple of these labels:
```
Comment (majakurcius) [inline collaboration/sig-core/proposals/test-execution-using-labels.md]:
```suggestion
Each `TestDefinition` should be labeled with one or more of these labels:
```

File: collaboration/sig-core/proposals/test-execution-using-labels.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+* `kyma-project.io/test.before-backup=true`
+* `kyma-project.io/test.after-restore=true`
+* `kyma-project.io/test.before-upgrade=true`
+* `kyma-project.io/test.after-upgrade=true`
+* `kyma-project.io/test.integration=true`
+
+>*This list can be extended at will.*
```
Comment (majakurcius) [inline collaboration/sig-core/proposals/test-execution-using-labels.md]:
```suggestion
> **NOTE:** This list can be extended at will.
```

File: collaboration/sig-core/proposals/test-execution-using-labels.md [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+Shell scripts install `ClusterTestSuites` with corresponding label queries e.g.:
```
Comment (majakurcius) [inline collaboration/sig-core/proposals/test-execution-using-labels.md]:
```suggestion
Shell scripts install `ClusterTestSuites` with corresponding label queries. For example:
```

File: collaboration/sig-core/proposals/test-execution-using-labels.md [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+  count: 1
+  selectors:
+    matchLabels:
+    - kyma-project.io/test.before-backup: true
+```
+
+
+## Implementation
+
+>*Octopus` does not require any changes*
```
Comment (majakurcius) [inline collaboration/sig-core/proposals/test-execution-using-labels.md]:
```suggestion
>**NOTE:** Octopus` does not require any changes.
```

File: collaboration/sig-core/proposals/test-execution-using-labels.md [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+
+* Kyma CLI
+  * list available scenarios
+  * support creation of `ClusterTestSuite` based on scenario(s)
+* Existing Tests:
+  * apply labels to all `TestDefinitions` including the ones that are not part of a basic Kyma installation (e.g. tests-folder)
+* CI-Scripts
+  * remove custom test execution logic and replace by kyma-cli call
+* Upgrade and Backup/Restore tests:
+  * get rid off old upgrade test framework / backup/restore test framework by refactoring the implemented tests into their own test definitions
```
Comment (majakurcius) [inline collaboration/sig-core/proposals/test-execution-using-labels.md]:
```suggestion
* Kyma CLI:
  * List available scenarios.
  * Support creation of `ClusterTestSuite` based on scenario(s).
* Existing Tests:
  * Apply labels to all `TestDefinitions`, including the ones that are not part of the basic Kyma installation (e.g. the `tests` folder).
* CI-Scripts:
  * Remove custom test execution logic and replace it with a `kyma-cli` call.
* Upgrade and Backup/Restore tests:
  * Get rid off old upgrade test framework/backup/restore test framework by refactoring the implemented tests into their own test definitions.
```

File: collaboration/sig-core/proposals/test-execution-using-labels.md [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+
+
+### Optional Changes
+
+* move additional `Helm-charts` from tests-folder into default installation (resources-folder)
```
Comment (majakurcius) [inline collaboration/sig-core/proposals/test-execution-using-labels.md]:
```suggestion
* Move additional Helm charts from the `tests` folder into the default installation (the `resources` folder).
```

File: collaboration/sig-core/proposals/test-execution-using-labels.md [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+  >*introduce additional/optional component 'additional-tests'*
```
Comment (majakurcius) [inline collaboration/sig-core/proposals/test-execution-using-labels.md]:
```suggestion
  * Introduce an additional/optional component named `additional-tests`.
```
Is this 
1. another optional change, or 
2. a way you can move those Helm charts you talk about in the line above? 

If `1.`, please merge that with the sentence above and get rid of the italics. 
If `2.`, please make it another bullet point and get rid of the italics (see the suggestion).

Comment (k15r) [reply collaboration/sig-core/proposals/test-execution-using-labels.md]: its one way to add all those charts to the resources folder. 
I dont know if there is a better approach to do so so

File: collaboration/sig-core/proposals/test-execution-using-labels.md [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+  count: 1
+  selectors:
+    matchLabels:
+    - kyma-project.io/test.before-backup: true
+```
+
+
+## Implementation
+
+>**NOTE:** Octopus` does not require any changes.
```
Comment (majakurcius) [inline collaboration/sig-core/proposals/test-execution-using-labels.md]:
```suggestion
>**NOTE:** Octopus does not require any changes.
```

File: collaboration/sig-core/proposals/test-execution-using-labels.md [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+* CI-Scripts:
+  * Remove custom test execution logic and replace it with a `kyma-cli` call.
+* Upgrade and Backup/Restore tests:
+  * Get rid off old upgrade test framework/backup/restore test framework by refactoring the implemented tests into their own test definitions.
+
+
+### Optional Changes
+
+* Move additional Helm charts from the `tests` folder into the default installation (the `resources` folder).
+  >*introduce additional/optional component 'additional-tests'*
```
Comment (majakurcius) [inline collaboration/sig-core/proposals/test-execution-using-labels.md]:
```suggestion
* Move additional Helm charts from the `tests` folder into the default installation (the `resources` folder). For example, introduce an additional/optional `additional-tests` component.
```

File: collaboration/sig-core/proposals/test-execution-using-labels.md (hunk 15 above)
Comment (majakurcius) [reply collaboration/sig-core/proposals/test-execution-using-labels.md]: I see. In that case, I'd go with `1.`. I've made a suggestion.

**Pull request merged**